snowshu/adapters/target_adapters/base_target_adapter.py

```diff
@@ -152,156 +105,90 @@ def load_data_into_relation(self, relation: Relation, data: pd.DataFrame) -> Non
-                                     relations: Iterable['Relation']) -> None:
-        """Applies all available source functions to target.
-
-        Looks for a function sql file in ./functions, executes against target for each db if it is.
+    def create_function_if_available(
+        self, function: str, relations: Iterable["Relation"]
+    ) -> None:
+        """Applies all available source functions to target. Looks for a function
+            sql file in ./functions, executes against target for each db if it is.
 
         Args:
             function: The name of the function, must match the sql file name exactly.
             relations: An iterable of relations to apply the function to.
         """
         try:
-            functions_path = os.path.abspath(os.path.join(
-                os.path.dirname(__file__),
-                self.name + '_adapter',
-                'functions'))
-            with open(os.path.join(functions_path, f'{function}.sql'), 'r') as function_file:  # noqa pylint: disable=unspecified-encoding
-                function_sql = function_file.read()
+            functions_path = (
+                Path(__file__).resolve().parent / f"{self.name}_adapter" / "functions"
+            )
+            function_sql = (functions_path / f"{function}.sql").read_text()
 
-            unique_schemas = {(rel.database, rel.schema,) for rel in relations}
+            unique_schemas = {
+                (
+                    rel.database,
+                    rel.schema,
+                )
+                for rel in relations
+            }
             for db, schema in unique_schemas:  # noqa pylint: disable=invalid-name
                 database = self._correct_case(db)
                 schema = self._correct_case(schema)
-                conn = self.get_connection(database_override=database,
-                                           schema_override=schema)
-                logger.debug('Applying function %s to "%s"."%s"...', function, db, schema)
+                conn = self.get_connection(
+                    database_override=database, schema_override=schema
+                )
+                logger.debug(
+                    'Applying function %s to "%s"."%s"...', function, db, schema
+                )
                 conn.execute(function_sql)
-                logger.debug('Function %s added.', function)
+                logger.debug("Function %s added.", function)
         except FileNotFoundError:
-            logger.info('Function %s is not implemented for target %s.', function, self.CLASSNAME)
+            logger.info(
+                "Function %s is not implemented for target %s.",
+                function,
+                self.CLASSNAME,
+            )
+        except Exception as exc:
+            logger.error(
+                "Error applying function %s to target %s: %s",
+                function,
+                self.CLASSNAME,
+                exc,
+            )
+            raise
```

Comment:
should we change it `raise from` to provide more context about why the exception occurred?

Reply:
I agree